### PR TITLE
Fix Blazor WASM SDK packaging and backport fix

### PR DESCRIPTION
**Pull request by captainsafia**
**Changes in this PR**

This PR has a couple of changes to get the dev SDK hooked up correctly and to backport a fix from the ASP.NET Core repo. Below is a guideline for the changes in the PR.

**Remove BlazorWebAssembly SDK from list of bundled SDKs to ensure the version in the repo is used not a globally installed one**
* https://github.com/dotnet/sdk/pull/14968/files#diff-1ea18ff65faa2ae6fed570b83747086d0317f5e4bc325064f6c14319a9c4ff67L125

**Add targets for Blazor WASM SDK to solution file**
* https://github.com/dotnet/sdk/pull/14968/files#diff-15adc48e2ec26a16c1b83f53f5e4583d904fc7480c6f30e9df5a520a2a790ef1R273-R280

**Fix references to tools and tasks in target file to match new layout**
* https://github.com/dotnet/sdk/pull/14968/files#diff-3383a3d4e6696d707e6714511fb7e58f85a7b58319d0743047713992351f3abbL38-L39

**Fixed up layout generation for tasks**
* https://github.com/dotnet/sdk/pull/14968/files#diff-77ceea1be0cfda2bad0d3ad55960d8584c4cd2c684045db9df196957bf64fbcaL1

**Made sure that `System.CommandLine` dependency was copied over into tool**
* https://github.com/dotnet/sdk/pull/14968/files#diff-84643103df5f8dd13e17fb3b72d2856e69119d80656c206a8d607a5af926ef7eR193

**Ensure the Blazor WASM tool is published before the SDK**
* https://github.com/dotnet/sdk/pull/14968/files#diff-84643103df5f8dd13e17fb3b72d2856e69119d80656c206a8d607a5af926ef7eR184

**Update test project file to ensure that project-level test targets can run**
* https://github.com/dotnet/sdk/pull/14968/files#diff-bc8143061fec1517c068def5773a5325d95a226c065160c59e24ddb3e6cb16ecR1

**Backport fix from https://github.com/dotnet/aspnetcore/pull/28539 to the SDK here**
* https://github.com/dotnet/sdk/pull/14968/files#diff-3383a3d4e6696d707e6714511fb7e58f85a7b58319d0743047713992351f3abbR466
* https://github.com/dotnet/sdk/pull/14968/files#diff-c4ff3e90606b3bcda3cd69e7cf0e61c85618bdc53fbb8e4e61065e86261e42aaR249

### CODE REVIEW

File: src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj
Comment (wli3) [inline src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj:2]: why this change is needed?

File: src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj
Comment (wli3) [inline src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj:7]: why you need to customize output for a test?

File: src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj:2]: I was following the convention in some of [the](https://github.com/dotnet/sdk/blob/83d9d6971af6cdf68e409f7f9baa5501093f5018/src/Tests/Microsoft.NET.Restore.Tests/Microsoft.NET.Restore.Tests.csproj) [other](https://github.com/dotnet/sdk/blob/83d9d6971af6cdf68e409f7f9baa5501093f5018/src/Tests/Microsoft.NET.Build.Tests/Microsoft.NET.Build.Tests.csproj) [test](https://github.com/dotnet/sdk/blob/83d9d6971af6cdf68e409f7f9baa5501093f5018/src/Tests/Microsoft.NET.Pack.Tests/Microsoft.NET.Pack.Tests.csproj) [project](https://github.com/dotnet/sdk/blob/83d9d6971af6cdf68e409f7f9baa5501093f5018/src/Tests/Microsoft.NET.ToolPack.Tests/Microsoft.NET.ToolPack.Tests.csproj) [files](https://github.com/dotnet/sdk/blob/83d9d6971af6cdf68e409f7f9baa5501093f5018/src/Tests/Microsoft.NET.Clean.Tests/Microsoft.NET.Clean.Tests.csproj). I noticed that some projects use `<Project Sdk="">` directly. I believe that import enforces importing the props file first and the targets file last so it shouldn't be functionally different from what is here.

File: src/BlazorWasmSdk/Tasks/BrotliCompress.cs
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/BrotliCompress.cs:89]: lol

File: src/BlazorWasmSdk/Tool/Program.cs
Comment (pranavkm) [inline src/BlazorWasmSdk/Tool/Program.cs:38]: Why do we need this change? The previous variables looked a lot better.

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:76]: Could we format this across multiple lines?

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: Is there a reason this changed from `net472` -> `net46`?

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:9]: Could you add explain why this is done this way? Also if you could add a comment to that effect that would be great

File: src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets
Comment (pranavkm) [inline src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets:39]: Tool http://www.toolbandprojects.com/wp-content/uploads/2010/03/Tool-Band-Picture.jpg

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:42]:
```suggestion
    <AdditionalContent Include="$(BlazorWasmSdkRoot)Targets\**\*.*">
```

Friends don't let friends use double slashes.

**Comment by wli3** [comment]
also could you update the CODEOWNERS. Instead of your name, replace it with your team's alias?

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:46]:
```suggestion
    <AdditionalContent Include="$(BlazorWasmSdkRoot)Sdk\**\*.*">
```

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [inline src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:60]: Why is this changing?

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:9]: I guess it's a common theme across this repo. Let's ignore this.

Comment (wli3) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:9]: some of the tests are "exported" as a dotnet tools. These tests are not, so we don't need the extra complication.

File: src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj
Comment (captainsafia) [reply src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests.csproj]: Thought I needed this but turns out it is not the case. Reverted.

File: src/BlazorWasmSdk/Tool/Program.cs
Comment (captainsafia) [reply src/BlazorWasmSdk/Tool/Program.cs:38]: The `System.CommandLine` API matches variable names passed on the string you pass for the option in the `new Option` call. For example, if you create an option `new Option<string>('--intOptions')` then the variable name in the handler would be `intOptions`. An alternative solution to this would be to update the tool to take `--sources` and `--outputs` as options, update the Brotli compression task, and then use `sources` and `outputs` as variable names here. I figured this was the smallest delta.

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (captainsafia) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj]: Done.

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (captainsafia) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: It matches the target framework in the [original SDK](https://github.com/dotnet/aspnetcore/blob/f7e1712db88db4c01fcc49e5995d9528dbf608df/src/Components/WebAssembly/Sdk/src/Microsoft.NET.Sdk.BlazorWebAssembly.csproj#L4).

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (captainsafia) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:60]: These packages are downgrade to versions that support the `net46` target framework mentioned above.

File: src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj
Comment (pranavkm) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: Could we change this to `net472` if it lets us use `MicrosoftBuildUtilitiesCoreVersion`?

Comment (captainsafia) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: Tried to get this working. Changing the target to net472 breaks the full framework tests which target net48 by default. I updated the target framework for the tests [here](https://github.com/dotnet/sdk/pull/14968/commits/52f0752688c01b8914038622f3f5f6b97933db99) but the end result is that it tries to build for both net472 and .NET Core on all Windows tests, including the ones that don't validate the full framework.

@wli3 Is there a way to configure the framework to target net472 and only build the tests for net472 under the full framework tests on Windows and not the other builds on Windows?

Comment (wli3) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: not other full frameworks like net48 or not on .net core?

Comment (captainsafia) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: Correct. Assuming you don't explicitly set the TargetFramework in the Test project, the tests will be built to target .NET Core. In the full framework scenario, the tests will be built to target .NET Core, but the underlying MSBuild will validate on .NET Framework (at least that's what I've assessed from the current test run). Is there a way to build the tests to target .NET Core, but to use a particular framework version in the evaluation?

Comment (pranavkm) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: If building with net46 requires pinning the tasks versions, but does not have any other side-effects (all the tests appear to be passing), can we check that in and file follow up issue to clean this up? I wouldn't want this PR to languish for several weeks because of this. Particularly because it blocks VS \ James from doing their gRPC work. (FYI @mkArtakMSFT)

Comment (mkArtakMSFT) [reply src/BlazorWasmSdk/Tasks/Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.csproj:38]: Yes, we can do that. As long as the tests don't fail, we should be good to merge this and follow-up later.

**Comment by captainsafia** [comment]
@wli3 Post-holiday bump on this PR.

**Review by wli3 — APPROVED** [review]

**Comment by wli3** [comment]
@captainsafia also need an approval from your team member

**Comment by wli3** [comment]
@captainsafia and from now on, if you don't think a change to Blazor WASM would affect SDK, you don't need my team's approval. Since these code belong to your team in the end.

**Review by pranavkm — APPROVED** [review]

**Pull request merged**
